python/tvm/contrib/target/coreml.py

```diff
@@ -0,0 +1,252 @@
+        self.out_map[const] = [output]
+
+    def visit_var(self, var):
+        name = var.name_hint
+        shape = [int(n) for n in var.type_annotation.shape]
+        dtype = var.type_annotation.dtype
+        self.model_inputs_.append((name, shape, dtype))
+        self.out_map[var] = [name]
+
+    def visit_call(self, call):
+        inputs = []
+        for arg in call.args:
+            super().visit(arg)
+            for out in self.out_map[arg]:
+                inputs.append(out)
+        outputs = ["buf_" + str(self.buf_idx_)]
+        op_name = call.op.name
+        layer_name = op_name + "_" + str(self.buf_idx_)
+
+        assert op_name in _convert_map, "{} is not supported".format(op_name)
+        _convert_map[op_name](self.builder, layer_name, inputs, outputs,
+                              call.args, call.attrs)
+
+        self.buf_idx_ = self.buf_idx_ + 1
+        self.out_map[call] = outputs
+
+    def compile(self, out_dir):
+        """
+        Build a Core ML model and compile it with Xcode toolchain.
+        """
+        import coremltools
+        from coremltools.proto.Model_pb2 import ArrayFeatureType
+
+        FEATURE_TYPE_MAP = {
+            "float32": ArrayFeatureType.FLOAT32,
+            "float64": ArrayFeatureType.DOUBLE,
+            "int32": ArrayFeatureType.INT32,
+        }
+
+        input_names, input_dims, input_dtypes = zip(*self.model_inputs_)
+        self.builder.set_input(input_names, input_dims)
+        for i, dtype in enumerate(input_dtypes):
+            assert dtype in FEATURE_TYPE_MAP
+            input_desc = self.builder.spec.description.input
+            input_desc[i].type.multiArrayType.dataType = FEATURE_TYPE_MAP[dtype]
+
+        output_dim = [int(n) for n in self.function.ret_type.shape]
+        self.builder.set_output(self.out_map[self.function.body], [output_dim])
+        for i, dtype in enumerate([self.function.ret_type.dtype]):
+            assert dtype in FEATURE_TYPE_MAP
+            output_desc = self.builder.spec.description.output
+            output_desc[i].type.multiArrayType.dataType = FEATURE_TYPE_MAP[dtype]
+
+        model = coremltools.models.MLModel(self.builder.spec)
+        xcode.compile_coreml(model, self.model_name, out_dir)
+
+
+@tvm._ffi.register_func("relay.ext.coremlcompiler")
+def coreml_compiler(ref):
+    """
+    Create a CoreML runtime from a Relay module.
+    """
+    model_dir = os.getcwd()
+    if isinstance(ref, tvm.ir.module.IRModule):
+        for var, func in ref.functions.items():
+            name = var.name_hint
+            builder = CodegenCoreML(name, func)
+            builder.visit(func.body)
+            mlmodelc_path = "{}/{}.mlmodelc".format(model_dir, name)
+            if os.path.exists(mlmodelc_path):
+                shutil.rmtree(mlmodelc_path)
+            builder.compile(model_dir)
+
+    ctx = tvm.cpu(0)
+    return coreml_runtime.create(model_dir, ctx).module
+
+
+def _register_coreml_op(op_name):
```

Comment:
This should go in python/tvm/relay/op/contrib/coreml.py (you can take a look at dnnl.py in the same directory).